// redux-wordpress/lib/reducer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = createReducer;

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Reducer = function () {
	function Reducer(name) {
		_classCallCheck(this, Reducer);

		this.name = name;
	}

	_createClass(Reducer, [{
		key: 'match',
		value: function match() {
			return false;
		}
	}, {
		key: 'map',
		value: function map(state) {
			return state;
		}
	}, {
		key: 'prepareNewState',
		value: function prepareNewState(action) {
			var state = {};

			if (action.results) {
				state.data = action.results;
			} else if (action.result) {
				state.data = action.result;
			}

			if (typeof action.total !== 'undefined') {
				state.total = action.total;
			}

			if (typeof action.totalPages !== 'undefined') {
				state.totalPages = action.totalPages;
			}

			return state;
		}
	}]);

	return Reducer;
}();

var FetchReducer = function (_Reducer) {
	_inherits(FetchReducer, _Reducer);

	function FetchReducer() {
		_classCallCheck(this, FetchReducer);

		return _possibleConstructorReturn(this, (FetchReducer.__proto__ || Object.getPrototypeOf(FetchReducer)).apply(this, arguments));
	}

	_createClass(FetchReducer, [{
		key: 'match',
		value: function match(type) {
			return type.match(new RegExp('^@@wp/' + this.name + '/fetched/(\\w+)$'));
		}
	}, {
		key: 'map',
		value: function map(state, action) {
			var obj = {};
			var self = this;
			var match = self.match(action.type);

			obj[match[1]] = Object.assign({}, state[match[1]] || {}, self.prepareNewState(action));

			return Object.assign({}, state, obj);
		}
	}]);

	return FetchReducer;
}(Reducer);

var FetchAllReducer = function (_FetchReducer) {
	_inherits(FetchAllReducer, _FetchReducer);

	function FetchAllReducer() {
		_classCallCheck(this, FetchAllReducer);

		return _possibleConstructorReturn(this, (FetchAllReducer.__proto__ || Object.getPrototypeOf(FetchAllReducer)).apply(this, arguments));
	}

	_createClass(FetchAllReducer, [{
		key: 'match',
		value: function match(type) {
			return type.match(new RegExp('^@@wp/' + this.name + '/fetched-all/(\\w+)$'));
		}
	}]);

	return FetchAllReducer;
}(FetchReducer);

var FetchByIdReducer = function (_Reducer2) {
	_inherits(FetchByIdReducer, _Reducer2);

	function FetchByIdReducer() {
		_classCallCheck(this, FetchByIdReducer);

		return _possibleConstructorReturn(this, (FetchByIdReducer.__proto__ || Object.getPrototypeOf(FetchByIdReducer)).apply(this, arguments));
	}

	_createClass(FetchByIdReducer, [{
		key: 'match',
		value: function match(type) {
			return type.match(new RegExp('^@@wp/' + this.name + '/fetched-by-id/(\\w+)$'));
		}
	}, {
		key: 'map',
		value: function map(state, action) {
			var obj = {};
			var match = this.match(action.type);

			obj[match[1] + '/' + action.id] = this.prepareNewState(action);

			return Object.assign({}, state, obj);
		}
	}]);

	return FetchByIdReducer;
}(Reducer);

var FetchEndpointReducer = function (_Reducer3) {
	_inherits(FetchEndpointReducer, _Reducer3);

	function FetchEndpointReducer() {
		_classCallCheck(this, FetchEndpointReducer);

		return _possibleConstructorReturn(this, (FetchEndpointReducer.__proto__ || Object.getPrototypeOf(FetchEndpointReducer)).apply(this, arguments));
	}

	_createClass(FetchEndpointReducer, [{
		key: 'match',
		value: function match(type) {
			return type.match(new RegExp('^@@wp/' + this.name + '/fetched/(\\w+)/(\\w+)$'));
		}
	}, {
		key: 'map',
		value: function map(state, action) {
			var obj = {};
			var self = this;
			var match = self.match(action.type);

			obj[match[1] + '/' + match[2]] = self.prepareNewState(action);

			return Object.assign({}, state, obj);
		}
	}]);

	return FetchEndpointReducer;
}(Reducer);

var FetchEndpointByIdReducer = function (_Reducer4) {
	_inherits(FetchEndpointByIdReducer, _Reducer4);

	function FetchEndpointByIdReducer() {
		_classCallCheck(this, FetchEndpointByIdReducer);

		return _possibleConstructorReturn(this, (FetchEndpointByIdReducer.__proto__ || Object.getPrototypeOf(FetchEndpointByIdReducer)).apply(this, arguments));
	}

	_createClass(FetchEndpointByIdReducer, [{
		key: 'match',
		value: function match(type) {
			return type.match(new RegExp('^@@wp/' + this.name + '/fetched-by-id/(\\w+)/(\\w+)$'));
		}
	}, {
		key: 'map',
		value: function map(state, action) {
			var obj = {};
			var self = this;
			var match = self.match(action.type);

			obj[match[1] + '/' + action.id + '/' + match[2]] = self.prepareNewState(action);

			return Object.assign({}, state, obj);
		}
	}]);

	return FetchEndpointByIdReducer;
}(Reducer);

var FetchAllEndpointReducer = function (_FetchEndpointReducer) {
	_inherits(FetchAllEndpointReducer, _FetchEndpointReducer);

	function FetchAllEndpointReducer() {
		_classCallCheck(this, FetchAllEndpointReducer);

		return _possibleConstructorReturn(this, (FetchAllEndpointReducer.__proto__ || Object.getPrototypeOf(FetchAllEndpointReducer)).apply(this, arguments));
	}

	_createClass(FetchAllEndpointReducer, [{
		key: 'match',
		value: function match(type) {
			return type.match(new RegExp('^@@wp/' + this.name + '/fetched-all/(\\w+)/(\\w+)$'));
		}
	}]);

	return FetchAllEndpointReducer;
}(FetchEndpointReducer);

var FetchAllEndpointByIdReducer = function (_FetchEndpointByIdRed) {
	_inherits(FetchAllEndpointByIdReducer, _FetchEndpointByIdRed);

	function FetchAllEndpointByIdReducer() {
		_classCallCheck(this, FetchAllEndpointByIdReducer);

		return _possibleConstructorReturn(this, (FetchAllEndpointByIdReducer.__proto__ || Object.getPrototypeOf(FetchAllEndpointByIdReducer)).apply(this, arguments));
	}

	_createClass(FetchAllEndpointByIdReducer, [{
		key: 'match',
		value: function match(type) {
			return type.match(new RegExp('^@@wp/' + this.name + '/fetched-all-by-id/(\\w+)/(\\w+)$'));
		}
	}]);

	return FetchAllEndpointByIdReducer;
}(FetchEndpointByIdReducer);

function createReducer(name) {
	var reducers = [new FetchReducer(name), new FetchAllReducer(name), new FetchByIdReducer(name), new FetchEndpointReducer(name), new FetchEndpointByIdReducer(name), new FetchAllEndpointReducer(name), new FetchAllEndpointByIdReducer(name)];

	return function () {
		var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
		var action = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

		if (!action.ok) {
			return state;
		}

		var newState = Object.assign({}, state);

		reducers.forEach(function (reducer) {
			if (reducer.match(action.type)) {
				newState = reducer.map(newState, action);
			}
		});

		return newState;
	};
}
